src/eval/Eval.ml

```diff
@@ -572,7 +572,7 @@ let create_cur_state_fields initcstate curcstate =
             ~f:(fun (t, li) ->
               let%bind t1 = fromR @@ literal_type lc in
               let%bind t2 = fromR @@ literal_type li in
-              if s = t && type_equiv t1 t2 then pure true else fail0 "")
+              if s = t && type_assignable t1 t2 then pure true else fail0 "")
```

Comment:
This code is asserting that the contract state as obtained from `state.json` (`t1`) is a superset of the state obtained from contract field initializers (`t2`). I think the operands of `type_assignable` must be reversed.

Reply:
Actually, this should still be `type_equiv`. The types are derived from the literal values in the code and the JSON, and there is no way to construct a literal value with an Address type, as byte string literals are always considered of type ByStr20.

If a field initialiser contains a byte string literal, then the intention is to consider that a ByStr20 until the point where we update the field environment with the value of the initialiser. At that point we check whether the field is declared to have an Address type, in which case we perform an IPC call to check that the byte string points to an address containing a contract with fields of the appropriate types.